logstash-core/src/main/java/org/logstash/secret/SecretIdentifier.java

```diff
@@ -0,0 +1,101 @@
+    private final static Pattern urnPattern = Pattern.compile("urn:logstash:secret:v1:.*$");
+    private final String VERSION = "v1";
+    private final String key;
+
+    /**
+     * Constructor
+     *
+     * @param key The unique part of the identifier. This is the key to reference the secret, and the key itself should not be sensitive. For example: {@code db.pass}
+     */
+    public SecretIdentifier(String key) {
+        this.key = validateWithTransform(key, "key");
+    }
+
+    /**
+     * Converts an external URN format to a {@link SecretIdentifier} object.
+     *
+     * @param urn The {@link String} formatted identifier obtained originally from {@link SecretIdentifier#toExternalForm()}
+     * @return The {@link SecretIdentifier} object used to identify secrets, null if not valid external form.
+     */
+    public static SecretIdentifier fromExternalForm(String urn) {
+        if (urn == null || !urnPattern.matcher(urn).matches()) {
+            return null;
+        }
+        String[] parts = colonPattern.split(urn, 5);
+        return new SecretIdentifier(validateWithTransform(parts[4], "key"));
+    }
+
+    /**
+     * Minor validation and transformation on input. Converts ":" to "_" to avoid URN conflicts, and downcases the parts
+     *
+     * @param part     The part of the URN to validate
+     * @param partName The name of the part used for logging.
+     * @return The validated and transformed part.
+     */
+    private static String validateWithTransform(String part, String partName) {
+        if (part == null || part.isEmpty()) {
+            throw new IllegalArgumentException(String.format("%s may not be null or empty", partName));
+        }
+        return part.replace(":", "_").toLowerCase(Locale.US);
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+
+        SecretIdentifier that = (SecretIdentifier) o;
+
+        if (key != null ? !key.equals(that.key) : that.key != null) return false;
+        return VERSION != null ? VERSION.equals(that.VERSION) : that.VERSION == null;
+    }
+
+    /**
+     * Get the key (unique part) of the identifier
+     *
+     * @return the unique part of the identifier
+     */
+    public String getKey() {
+        return key;
+    }
+
+    @Override
+    public int hashCode() {
+        int result = key != null ? key.hashCode() : 0;
+        result = 31 * result + (VERSION != null ? VERSION.hashCode() : 0);
+        return result;
+    }
+
+    /**
+     * Converts this object to a format acceptable external {@link String} format. Note - no gauruntees are made with respect to encoding or safe use. For example, the external
+     * format may not be URL safely encoded.
+     *
+     * @return the externally formatted {@link String}
+     */
+    public String toExternalForm() {
+        StringBuilder sb = new StringBuilder(100);
```

Comment:
Thoughts on using String.format instead?

`String.format("urn:logstash:secret:%s:%s", VERSION, this.key == null ? "-" : this.key.toLowerCase(Locale.US))`

Feels shorter, though maybe you have reasons for the StringBuilder use?

Reply:
It's just a minor style preference, imo it is easier to read as is.

Reply:
Sounds ok to me -- I have no strong opinion and the method functions as intended. Thanks for letting me into your brain a bit :)